process/HOPE-3--Dependency-Injection.md

```diff
@@ -0,0 +1,169 @@
+
+
+@hug.get()
+def my_endpoint(timer: hug.directives.timer):
+    pass
+```
+
+
+### Defining a dependency
+
+Defining a dependency will happen via a new hug `provides` decorator:
+
+```python
+    @hug.provides()
+    def mysql_connection(request, config):
+        ...
+```
+
+By default, the name will be the name of the function, but the first argument sent to that directive should enable redefining it:
+
+```python
+    @hug.provides("mysql")
+    def mysql_connection(request, config):
+        ...
+```
+
+It should also be possible to specify different dependency providers for different interfaces:
+
+```python
+    # Multiple interfaces using one provider
+    @hug.provides("mysql", interfaces=['http', 'websocket'])
+    def mysql_web():
+        ...
+
+
+    # For a provider for a single interface, this shorthand can be used
+    @hug.cli.provides("mysql")
+    def mysql_cli():
+        ...
+
+
+    # The following would also be a valid way to apply to multiple interfaces
+    @hug.http.provides("mysql")
+    @hug.cli.provides("mysql")
+    def mysql():
+        ...
+```
+
+### Using a dependency
+
+Dependencies will be pulled into individual endpoints via a `requires(` function, that will be called within function/method defaults:
+
+```python
+from hug import requires
+
+@hug.http()
+def hello_world(mysql=requires('mysql')):
+    pass
+```
+
+Arguments can be passed to the `requires` function, these will be passed directly to the `provides` function.
+This will be done in the same manner as Python's `funtools.partial` and follow the same function signature.
+Any non-provided arguments will then require the dependency to pull them from the current application action,
+like for normal hug calls, or via dependency injection. A key point: Any dependency can have unlimited sub-dependencies.
+
+Optionaly, you can also directly require any other normal Python callable:
+
+```python
+import json
+
+from hug import requires
+
+
+def shared_configuration(config_file_location="config_file.json"):
+    with open(config_file_location) as config_file:
+         return json.loads(config_file.read())
+
+
+@hug.http()
+def hello_world(config=requires(shared_configuration)):
+    pass
+```
+
+### Where registered dependency providers are stored
+
+Dependencies providers will be stored within the Hug API module level singleton, within a `dependency_providers` dictionary.
+If a second `provides` function is defined by default it will raise an `ExistingDependency` exception. However, if that second definition defines `override=True`, it will simply take the place of the first - similar to defining a second function with the same name. If a dependency provider is only meant for a particular set of interfaces, it will be stored on the sub interface API object within the singleton. Hug will first check the per-interface dict of providers, and only fall through to the global dict if the interface specific one does not have the requested provider.
```

Comment:
Should there be any type of warning when defining an interface local `provides` that will shadow a global `provides` of the same name? Alternatively should there be a restriction that you can either have a global `provides` or interface local `provides` for a given name?

Reply:
These are great questions! I can definitely see a case where it might make sense to have a default depends and an override for just one interface, but I can also see that being done just as easily on accident. At first I was thinking we could overload `override=True` for any case where there is a global interface to allow creating an interface specific provider, but upon some more thought, I'm thinking that might also lead to unexpected errors as you may want to override the global provider for this interface but not a previous provider defined for it. What if we added an `override_global` parameter only to the per interface provider functions?